oomph/parser.oomph

```diff
@@ -120,44 +135,57 @@ class Parser(List[tokenizer::Token] tokens):
             return i
         return null
 
-    meth eliminate_operators(List[PrecedenceItem] magic_list, List[Str] operators):
+    meth eliminate_operators(List[PrecedenceItem] magic_list, List[Str] unary, List[Str] binary):
         while magic_list.length() >= 2:
-            switch magic_list.first():
-                case Str op:
-                    if util::list_contains(operators, op):
-                        # Unary operation
-                        let expr = get_expr(magic_list.get(1))
-                        delete_slice_from_list(magic_list, 0, 2)
-                        insert_to_list(magic_list, 0, new ast::UnaryOperator(op, expr))
-                        continue
-                case *:
-                    pass
-
-            let i_or_null = self.find_binary_operator(magic_list, operators)
-            if i_or_null != null:
-                let i = i_or_null.get()
+            let un_or_null = self.find_unary_operator(magic_list, unary)
+            let bin_or_null = self.find_binary_operator(magic_list, binary)
+
+            if un_or_null != null and bin_or_null != null:
+                if un_or_null.get() < bin_or_null.get():
+                    bin_or_null = null
+                elif un_or_null.get() > bin_or_null.get():
+                    un_or_null = null
+                else:
+                    # TODO: error
+                    print("ambiguous operators")
+                    assert(false)
+
+            if un_or_null != null:
+                let i = un_or_null.get()
+                let op_and_expr = delete_slice_from_list(magic_list, i, i+2)
+                let op = get_op(op_and_expr.first())
+                let expr = get_expr(op_and_expr.last())
+                insert_to_list(magic_list, i, new ast::UnaryOperator(op, expr))
+            elif bin_or_null != null:
+                let i = bin_or_null.get()
                 let lhs_op_rhs = delete_slice_from_list(magic_list, i, i+3)
                 let lhs = get_expr(lhs_op_rhs.first())
                 let op = get_op(lhs_op_rhs.get(1))
                 let rhs = get_expr(lhs_op_rhs.last())
                 insert_to_list(magic_list, i, new ast::BinaryOperator(lhs, op, rhs))
-                continue
+            else:
+                break
 
-            break
+    meth get_unary_operators() -> List[PrecedenceItem]:
+        let result = new List[PrecedenceItem]()
+        while self.peek().matches("keyword", "not") or self.peek().matches("op", "-"):
+            result.push(self.tokens.pop().value)
+        return result
 
     meth parse_expression() -> ast::Expression:
-        let ops = ["+", "-", "*", "/"]
+        let ops = ["+", "-", "*", "/", "and", "or", "not"]
```

Comment:
Rename `ops` to `binary_ops`, or get rid of it entirely? Should `"not"` be here?